README.md
# Knights Release 028 -- Source Code

This is the source code for KNIGHTS, a game of multi-player violent
dungeon bashing. (For more information about the game, please visit
https://www.knightsgame.org.uk/, or look in the `docs/` directory.)

The game was originally written for the Amiga by Kalle Marjola; this
PC version is by Stephen Thompson. (See the `amiga_knights` directory
for more about the original Amiga version.)

This is the open source version of the game. A Steam version is also
available (or will be soon). The Steam version is based on the same
code, but includes additional features enabled by the Steam platform,
such as matchmaking and friends invites.


## How to Build

For Windows build instructions please read `docs/building.html`.

For Linux, a Makefile is provided. This should work "out of the box"
on Debian, but it might require some tweaks for other distributions;
please read the comments at the top of the Makefile for further
information.


## Licence

Starting from release 026, Knights is licensed under the GNU General
Public Licence as published by the Free Software Foundation, either
version 2 of that Licence, or, at your option, any later version. See
`COPYRIGHT.txt` for full copyright details and `docs/GPL2.txt` and
`docs/GPL3.txt` for copies of the Licence itself. (The phrase "any
later version" is to be interpreted in accordance with Section 14 of
the GPL version 3.)

Please note that the above notice applies only to this open source
version of Knights. The Steam version of Knights is licensed
separately, and the GPL does not apply in that case.

---

Stephen Thompson \
stephen@solarflare.org.uk \
March 2026


docs/CHANGELOG.txt
Note: Numbers in square brackets refer to Trac
<http://www.knightsgame.org.uk:8000/>.


Release 021
===========

Released 10-Sep-2012

[#70, #94, #162, #183] New Lua scripting system added.

[#80] "Tutorial Mk II" added.

[#197] Changes made to bring gameplay closer to the original Knights:

  * Chance of healing potion reduced from 60% to 40%.

  * Chance of "Sense Items" scroll effect now increases steadily as
    the game progresses.

  * Creatures hit by crossbow bolts are now stunned for a short time.

[#142] (Scripting) Monster sound effects can now be defined on a
  per-monster basis.

[#161] (Scripting) ai_fear now accepts a list of items.

[#193] (Windows) Fixed full screen resolution issue (resolution
  changes weren't being detected until Knights was restarted).

[#184] (Windows) Support Unicode path names for knights_config.txt.

[#190] Fixed incorrect starting point in "Run for the Freedom" quest
  (when playing in single player mode).

[#201] Fixed bug: Quest requirements box was blank for observers.

[#186] "Gems Needed" message now displayed when striking book with
  wand.

[#188] "Time Limit" setting can no longer be changed by observers.



Release 020
===========

Released 7-Jul-2012

New features:

[#160] Team games can now be played for any quest type, not just Duel
  to the Death. (The House Colours determine whether a game is a team
  game or not; if two or more knights share a colour, then they play
  on the same team.) Also, five new House Colours were added.

[#170] Invisible knights are now shown with a transparency effect.
  Also, invisible knights can now be seen by other players on the same
  team.

[#28] Added a Quest Requirements display to the in-game screen. (This
  can be minimized, if you wish, by clicking on its title bar.)

[#159] The ` key (this is above TAB on most keyboards) can now be used
  as a shortcut to start typing a team chat message. Also, the chat
  keys can now be re-defined on the Options screen.

[#57] On Windows, Knights now uses DirectX 11 for drawing (if
  available). This results in lower CPU usage and (potentially)
  smoother animation. If DirectX is not available, or is not up to
  date enough, then SDL is used as a fallback (the Options screen will
  tell you whether this happened on your system).


Bug fixes:

[#155, #179] Fixed some dungeon generator bugs, including: switches in
  gnome room not always working; switches firing bolts out of the
  wrong square or in the wrong direction; inaccessible areas in some
  Ring dungeons.

[#175] When swinging an axe at the special pentagram, the axe was being
  swung, instead of thrown as usual. This has been fixed.

[#158] Fixed bug where you could sometimes walk through a flying axe
  or bolt if you were moving straight towards it.

[#167] If you set a trap on a chest that already had a dart trap, then
  the dart was firing out from the wrong square. Now fixed.

[#160] In Duel to the Death quests, if the black team used a wand of
  securing, the entry point was marked by a dark grey rectangle on a
  black background, which was very difficult to see. This has been
  changed to light grey.

[#119] Following some changes to the mini-map code (a couple of
  releases ago), the "closed gate" tiles had started appearing as
  walls on the mini-map. This has now been reverted.

[#100] Fixed bug where sounds were sometimes not played correctly (for
  example, if a knight in an adjacent room opened a door into your
  room, then the "door opening" sound was not being played.)

[#172] The in-game player list is now sorted so that players on the
  same team are grouped together.

[#151] The "team chat is available" message is no longer shown if you
  are the only player on your team.

[#77] Chat from the quest selection screen is now carried through into
  the game.


Other changes:

[#152] Zombies can no longer instantly respawn after they die. There
  is now a minimum time of five seconds after a zombie dies before it
  can respawn.

[#20] Added one more guarded exit room (with three locked doors
  instead of the usual two).

[#169] Knights now opens in windowed mode by default. (Full screen
  mode can be selected from the Options screen, as usual.)

[#130] Traceback added to Lua error messages. This should make it
  easier to debug any Lua related problems in the future.

[#171] "Frags" renamed to "points" in Deathmatch mode.

[#163] Pressing Enter with nothing typed will now exit chat mode
  (instead of doing nothing).



Release 019
===========

Released 8-Jan-2012

New features:

[#131] New "Deathmatch" quest type. Knights compete to score the
highest number of frags before the time runs out.

[#135] New entry point setting: "Different every time", in which your
entry point is randomized every time you respawn.

[#41] The dungeon generator can now randomly reflect and/or rotate the
predefined rooms. This introduces a bit more variety into the
generated dungeons.

[#128] Friendly fire is now disabled for team games. Note this applies
to direct weapon hits (e.g. sword hits or thrown daggers), but not
"indirect" hits (e.g. if you press a switch which fires a bolt into a
team mate, then they will still be harmed).

[#122] Team chat is now available.

[#123] More than two teams can play in team games.

[#121] Fixed bug where knights could not use an entry point if it had
been secured by another knight on the same team.

[#126] You can now map a room if a team mate is in it.

[#47] A message is displayed if the dungeon type had to be expanded
(this can happen e.g. if you choose a "Tiny" dungeon for a very large
quest).


Bug fixes and other improvements:

[#139] Fixed bug where zombies were walking down pits, and also
vampire bats were not being generated from pit tiles when they were
supposed to.


python/kconvert.py
#!/usr/bin/python

# Convert Knights KRD file into Kconfig format.

import sys


PRE = """# Room data, imported from KRD.
# This file was automatically generated by kconvert.py.

"""

POST = "eof"

facing=["'north'","'east'","'south'","'west'"]

def readbyte(f):
    return ord(f.read(1))

def gtrap(x, y, mjr, mnr):
    if mjr>=201 and mjr<=204:
        return ('shoot', x, y, mjr - 200)
    elif mjr == 220:
        return ('teleport', x, y)
    else:
        lo = min(mjr,mnr)
        hi = max(mjr,mnr)
        if (lo>=7 and lo<=10 and hi==lo+43) or (lo==16 and hi==64) \
        or (lo==17 and hi==65) or ((lo==47 or lo==49) and hi==lo+31):
            # door or pit toggle
            return ('toggle', x, y)
        elif ((hi >= 50 and hi<=53) or hi==64 or hi==65) and lo==hi:
            # door open
            return ('open', x, y)
        elif (hi==78 or hi==80) and lo==hi:
            # pit close
            return ('close', x, y)
        elif (hi==47 or hi==49) and lo==hi:
            # pit open
            return ('open', x, y)
        elif (hi==169 and lo==76):
            # pentagram toggle
            return ('toggle_no_sound', x, y)
        elif (hi==15 and lo==3):
            # crystal ball toggle
            return ('toggle', x, y)
        elif hi==15 and lo==15:
            # crystal ball open
            return ('open', x, y)
        elif hi==3 and lo==3:
            # crystal ball close
            return ('close', x, y)
        elif hi==2 and lo==2:
            return None
        else:
            print "bad trap"
            print x, y, mjr, mnr
            sys.exit(1)

def dowrite(f, t, xb, yb):
    f.write('%s %d %d %d' % (t[0], len(t)-1, t[1]-xb, t[2]-yb))
    if len(t)>3:
        f.write(' %d' % t[3])

def writetrap(f, tdat, xb, yb):
    t2 = []
    for t in tdat:
        if t != None:
            t2.append(t)

    if len(t2)==0: raise "Trap with no effect?"

    f.write('%d ' % len(t2))
    for i in range(len(t2)):
        dowrite(f, t2[i], xb, yb)
        if i != len(t2)-1:
            f.write(' ')

def matchtrap(tdat,x,y):
    for t in tdat:
        if t != None:
            if t[1]==x and t[2]==y: return True
    return False


# Main program.

# Check arguments
if len(sys.argv) != 3:
    print "usage:", sys.argv[0], "<infile> <outdir>"
    sys.exit(1)

# Open files
f = file(sys.argv[1], 'rb')

outdir = sys.argv[2] + '/'
outfiles = { 'standard_room':     file(outdir + 'standard_rooms.txt', 'w'),
             'gnome_room':        file(outdir + 'gnome_rooms.txt', 'w'),
             'guarded_exit':      file(outdir + 'guarded_exits.txt', 'w'),
             'liche_tomb':        file(outdir + 'liche_tombs.txt', 'w'),
             'special_pentagram': file(outdir + 'special_pentagrams.txt', 'w') }

for k in outfiles:
    outfiles[k].write(PRE+"\n")


# Skip AMOS header
# (we assume there is only one bank in the file)
f.read(20)


# Load numbers of rooms
na = readbyte(f)
nb = readbyte(f)
nc = readbyte(f)
nb = nb - nc
na = na - nb - nc
nx = readbyte(f)
nz = readbyte(f)
nrooms = na+nb+nc+nx+nz


# Load all rooms
for a in range(nrooms):
    mpp = a*1080+5 +20
    if a<nc:
        letter='C'
        nroom=a+1
    elif a<nb+nc:
        letter='B'
        nroom=a+1-nc
    elif a<na+nb+nc:
        letter='A'
        nroom=a+1-nc-nb
    elif a<na+nb+nc+nx:
        letter='X'
        nroom=a+1-nc-nb-na
    else:
        letter='Z'
        nroom=a+1-na-nb-nc-nx

    # Read layout information
    f.seek(mpp + 14*14*5)
    rums = readbyte(f)
    rumdata = []
    rumcheck = [[]]*14
    for i in range(14):
        rumcheck[i] = [0]*14
    f.seek(mpp + 14*14*5 + 4)
    for i in range(rums):
        x = readbyte(f)
        y = readbyte(f)
        sx = readbyte(f)
        sy = readbyte(f)
        rumdata.append([x,y,sx,sy])
        for xx in range(x, x+sx):
            for yy in range(y, y+sy):
                if (xx == x or xx == x+sx-1) and (yy == y or yy == y+sy-1): 
                    continue
                rumcheck[xx][yy] = rumcheck[xx][yy]+1
                if rumcheck[xx][yy] > 2:
                    print "ROOMS ERROR, more than 2 rooms at", xx, ",", yy
                    print rumdata
                    sys.exit(1)

    # Load traps table
    tx = []
    ty = []
    tmjr = []
    tmnr = []
    tad1 = []
    tad2 = []
    tdata = []
    
    while 1:
        trap_base = mpp + 14*14*5 + 52 + len(tx)*6
        f.seek(trap_base)
        x = readbyte(f)
        y = readbyte(f)
        mjr = readbyte(f)
        mnr = readbyte(f)
        ad1 = readbyte(f)
        ad2 = readbyte(f)

        if mjr == 0: break

        tx.append(x)
        ty.append(y)
        tmjr.append(mjr)
        tmnr.append(mnr)
        tad1.append(ad1)
        tad2.append(ad2)
        if len(tx)==8: break

    # Get traps data
    if len(tx)>0:
        for i in range(len(tx)):
            tdat=[]
            tdat.append(gtrap(tx[i], ty[i], tmjr[i], tmnr[i]))
            if tad1[i] > 0:
                j = tad1[i]-1
                tdat.append(gtrap(tx[j], ty[j], tmjr[j], tmnr[j]))
            if tad2[i] > 0: 
                j = tad2[i]-1
                tdat.append(gtrap(tx[j], ty[j], tmjr[j], tmnr[j]))
            tdata.append(tdat)

    # Read rooms themselves
    for nrot in range(4):
        bdd = nrot*14*14

        # work out type
        if letter=='A' or (letter=='X' and nrot>=2) or letter=='B' or letter=='C':
            roomtype = 'standard_room'
        elif (letter=='Z' and nrot==0) or (letter=='X' and nrot<=1):
            roomtype = 'guarded_exit'
        elif (letter=='Z' and nrot==1):
            roomtype = 'liche_tomb'
        elif (letter=='Z' and nrot==2):
            roomtype = 'gnome_room'
        elif (letter=='Z' and nrot==3):
            roomtype = 'special_pentagram'
        else:
            raise "Unknown room type"

        outfile = outfiles[roomtype]

        # Write room header
        outfile.write("segment\n")
        outfile.write('name: %s%02d.%d\n' % (letter, nroom, nrot+1))
        outfile.write('width: 12\n')
        outfile.write('height: 12\n')
        outfile.write('data:\n')
            
        # Read the map itself
        sx = []
        sy = []
        sn = []
        f.seek(mpp+bdd)
        r = []
        
        for y in range(14):
            for x in range(14):
                t = readbyte(f)  # tile number
                tr = 0
                if t>=191 and t<=199:
                    tr = t-190
                    t = 19 # switch down
                elif t>=181 and t<=189:
                    tr = t-180
                    t = 18 # switch up
                elif t>=170 and t<=179:
                    tr = t-170
                    t = 67 # pressure plate
                elif t>=160 and t<=168:
                    tr = t-160
                    t = 76 # pentagram
                elif t==169:
                    t = 1  # Live pentagram

                if (t == 1 or t == 76) and roomtype == "special_pentagram":
                    t = 86 #  Special pentagram

                if t == 73 and roomtype == "guarded_exit":
                    t = 91 #  Scratched floor tile + Vampire bat

                # Fix for Trac #20
                if (t >= 11 and t <= 14 and roomtype == "guarded_exit"
                    and not (letter == 'Z' and nroom == 9 and nrot == 0 and t == 11)):
                        t = t + 92 - 11        # convert to "Special home"

                if letter == "Z" and nroom == 3 and nrot == 0 and x == 12 and y == 10:
                    # This particular tile should be replaced with 81 (stair-top)
                    # to prevent the dungeon generator creating doors into the guarded exit area.
                    if t != 59:
                        print "Unexpected tile in room Z3:1"
                        print "Found:", t, "Expected:", 59
                        sys.exit(1)
                    t = 81
                        
                if t < 1 or (t > 95):
                    print "bad tile"
                    print x, y, t
                    raise "bad tile"

                if x==0 or x==13 or y==0 or y==13:
                    if t<>2:
                        raise "non wall tile on border"
                else:
                    if t == 43:
                        # horiz table -- may need gnome book or necronomicon
                        if roomtype == 'gnome_room':
                            t = 50
                        elif roomtype == 'liche_tomb':
                            t = 51

                r.append(t)

                if tr<>0:
                    sx.append(x)
                    sy.append(y)
                    sn.append(tr)
                    
        # (end of for loops over x,y)

        # Replace inaccessible iron doors with walls:
        def isok(x,y): 
            t = r[y*14+x]
            return x==0 or x==13 or y==0 or y==13 or t==9 or t==10 or t==2 or t==3

        for y in range(1,13):
            for x in range(1,13):
                result = \
                    isok(x-1, y-1) and \
                    isok(x-1, y  ) and \
                    isok(x-1, y+1) and \
                    isok(x,   y-1) and \
                    isok(x,   y  ) and \
                    isok(x,   y+1) and \
                    isok(x+1, y-1) and \
                    isok(x+1, y  ) and \
                    isok(x+1, y+1)
                if result: r[y*14+x] = 2

        # Now write the output
        for y in range(1,13):
            for x in range(1,13):
                tile = r[y*14 + x]

                if tile >= 7 and tile <= 10:
                    # Wood or iron door.
                    # These become "traplocked" if any switch targets this tile.
                    for i in range(len(sx)):
                        if matchtrap(tdata[sn[i]-1], x, y):
                            tile = tile + 80
                            break

                outfile.write("%2d " % tile)
            outfile.write("\n")

        # Do switches
        if (len(sx)>0):
            outfile.write('switches:\n')
            outfile.write(str(len(sx)))
            outfile.write('\n')
            for i in range(len(sx)):
                x=sx[i]
                y=sy[i]
                tr=sn[i]
                tile = r[y*14+x]
                outfile.write('%d %d ' % (x-1,y-1))
                writetrap(outfile, tdata[tr-1], x, y)
                outfile.write('\n')

        # Rooms data
        outfile.write('rooms:\n')
        outfile.write(str(len(rumdata)) + '\n')
        for rum in rumdata:
            for i in rum:
                outfile.write(str(i) + ' ')
            outfile.write('\n')

        # The end of the room
        outfile.write('end\n\n')

for k in outfiles:
    outfiles[k].write(POST+"\n")
    outfiles[k].close()
print 'wrote', nrooms, '* 4 rooms'
print na,nb,nc,nx,nz
